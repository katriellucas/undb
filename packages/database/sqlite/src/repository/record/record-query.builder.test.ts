import { EntityManager } from '@mikro-orm/better-sqlite'
import {
  createTestSelectField,
  createTestStringField,
  createTestTable,
  createTestView,
  TableSchema,
  Views,
  WithTableSchema,
  WithTableViews,
} from '@undb/core'
import { RecordSqliteQueryBuilder } from './record-query.builder'

const directions = ['asc', 'desc'] as const

describe('RecordSqliteQueryBuilder', () => {
  let queryBuilder: RecordSqliteQueryBuilder

  let em: EntityManager

  beforeAll(() => {
    // @ts-expect-error
    em = global.em
  })

  describe('sort', () => {
    describe('sort string field', () => {
      it.each(directions)('should order by string option %s', (direction) => {
        const table = createTestTable(
          new WithTableSchema(new TableSchema([createTestStringField({ id: 'fld1' })])),
          new WithTableViews(new Views([createTestView({ sorts: [{ fieldId: 'fld1', direction }] })])),
        )

        queryBuilder = new RecordSqliteQueryBuilder(em, table, null)
        queryBuilder.sort()

        const query = queryBuilder.qb.toQuery()

        expect(query).toMatchSnapshot()
      })
    })

    describe('sort select', () => {
      it.each(directions)('should order by select option %s', (direction) => {
        const table = createTestTable(
          new WithTableSchema(new TableSchema([createTestSelectField({ id: 'fld1' })])),
          new WithTableViews(new Views([createTestView({ sorts: [{ fieldId: 'fld1', direction }] })])),
        )

        queryBuilder = new RecordSqliteQueryBuilder(em, table, null)
        queryBuilder.sort()

        const query = queryBuilder.qb.toQuery()

        expect(query).toMatchSnapshot()
      })
    })
  })
})
