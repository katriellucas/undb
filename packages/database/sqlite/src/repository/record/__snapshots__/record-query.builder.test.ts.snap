// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RecordSqliteQueryBuilder > sort > sort select > should order by select option asc 1`] = `
"select * order by 
            CASE t.fld1
              WHEN 'opt1' THEN 0 
WHEN 'opt2' THEN 1 
              ELSE -1
            END
          "
`;

exports[`RecordSqliteQueryBuilder > sort > sort select > should order by select option desc 1`] = `
"select * order by 
            CASE t.fld1
              WHEN 'opt2' THEN 0 
WHEN 'opt1' THEN 1 
              ELSE 2
            END
          "
`;

exports[`RecordSqliteQueryBuilder > sort > sort string field > should order by string option asc 1`] = `"select * order by \`t\`.\`fld1\` asc"`;

exports[`RecordSqliteQueryBuilder > sort > sort string field > should order by string option desc 1`] = `"select * order by \`t\`.\`fld1\` desc"`;
